File: Profiler_Test/ProfilerTestBase.java
/*-----------------------------------------------------------------------+
 | eu.cqse.conqat.engine.testgap
 |                                                                       |
   $Id$
 |                                                                       |
 | Copyright (c)  2009-2015 CQSE GmbH                                 |
 +-----------------------------------------------------------------------*/
package eu.cqse.conqat.engine.testgap.profiler;

import static eu.cqse.conqat.engine.testgap.scope.traces.dotnet.DotNetTraceUtils.KEY_SEPARATOR;
import static eu.cqse.conqat.engine.testgap.scope.traces.dotnet.DotNetTraceUtils.LABEL_INLINED;
import static eu.cqse.conqat.engine.testgap.scope.traces.dotnet.DotNetTraceUtils.LABEL_JITTED;
import static eu.cqse.conqat.engine.testgap.scope.traces.dotnet.DotNetTraceUtils.keyValuesMapFor;
import static org.junit.Assert.assertEquals;
import static org.junit.Assume.assumeTrue;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import org.conqat.engine.core.bundle.BundleException;
import org.conqat.engine.core.logging.testutils.LoggerMock;
import org.conqat.engine.dotnet.ila.AssemblyExecutorTestBase;
import org.conqat.lib.commons.assertion.CCSMAssert;
import org.conqat.lib.commons.collections.CollectionUtils;
import org.conqat.lib.commons.collections.ListMap;
import org.conqat.lib.commons.filesystem.FileSystemUtils;
import org.conqat.lib.commons.io.ProcessUtils;
import org.conqat.lib.commons.io.ProcessUtils.ExecutionResult;
import org.conqat.lib.commons.string.StringUtils;
import org.conqat.lib.commons.system.SystemUtils;

import eu.cqse.conqat.engine.testgap.BundleContext;

/**
 * Base class for testing the .NET Profiler.
 */
public abstract class ProfilerTestBase extends AssemblyExecutorTestBase {

	/** Environment variable name to enable the profiler. */
	private static final String PROFILER_ENABLE_KEY = "COR_ENABLE_PROFILING";

	/** Environment variable name for the profiler's class ID. */
	private static final String PROFILER_CLASS_ID_KEY = "COR_PROFILER";

	/** Environment variable name for the directory to store profiler traces. */
	private static final String PROFILER_TARGETDIR_KEY = "COR_PROFILER_TARGETDIR";

	/** Environment variable name for the path to the profiler DLL. */
	private static final String PROFILER_PATH_KEY = "COR_PROFILER_PATH";

	/** Environment variable name to enable the profiler's light mode. */
	private static final String PROFILER_LIGHT_MODE_KEY = "COR_PROFILER_LIGHT_MODE";

	/** The profiler's class ID. */
	private static final String PROFILER_CLASS_ID = "{DD0A1BB6-11CE-11DD-8EE8-3F9E55D89593}";

	/** {@inheritDoc} */
	@Override
	public void initBundleContext() throws BundleException {
		new BundleContext(createBundleInfo(BundleContext.class));
	}

	/** Makes sure tests run only on Windows with .NET 4 and newer. */
	@Override
	public void setUp() throws BundleException, IOException {
		assumeTrue("Profiler only runs on Windows", SystemUtils.isWindows());
		super.setUp();
		boolean isDotNet4 = executeProgramInResourceFolder("netversioninfo.bat").getStdout()
				.contains("ProductVersion:   4.");
		assumeTrue("Profiler only runs on .NET 4 and newer", isDotNet4);
	}

	/** Sets all environment variables the profiler needs. */
	protected void registerProfiler(ProcessBuilder pb, File targetDir, boolean useLightMode, String bitness) {

		String profilerDll = "profiler-dotnet-newer/Profiler" + bitness + ".dll";

		// set environment variables for the profiler
		pb.environment().put(PROFILER_PATH_KEY, getAbsoluteResourcePath(profilerDll));
		pb.environment().put(PROFILER_TARGETDIR_KEY, targetDir.getAbsolutePath());
		pb.environment().put(PROFILER_CLASS_ID_KEY, PROFILER_CLASS_ID);
		pb.environment().put(PROFILER_ENABLE_KEY, "1");
		if (useLightMode) {
			pb.environment().put(PROFILER_LIGHT_MODE_KEY, "1");
		}
	}

	/**
	 * Asserts that the trace file written by the profiler has the same contents as
	 * the given reference trace, modulo some normalization.
	 */
	protected void assertNormalizedTraceFileEqualsReference(File referenceTraceFile, File directory,
			Set<String> assembliesToCompare) throws IOException {
		assertEquals("The normalized contents of the trace files did not match",
				getNormalizedTraceContent(referenceTraceFile, assembliesToCompare),
				getNormalizedTraceContent(getTraceFile(directory), assembliesToCompare));
	}

	/**
	 * Returns the trace.
	 *
	 * @throws AssertionError
	 *             if not exactly one trace was written.
	 */
	protected File getTraceFile(File directory) {
		File[] traceFiles = directory.listFiles((dir, name) -> name.startsWith("coverage_"));

		CCSMAssert.isTrue(traceFiles.length > 0, "No coverage trace was written.");
		CCSMAssert.isTrue(traceFiles.length <= 1,
				"More than one coverage trace was written: " + StringUtils.concat(traceFiles, ", "));

		return traceFiles[0];
	}

	/**
	 * Executes a program from the resource folder.
	 */
	protected static ExecutionResult executeProgramInResourceFolder(String program) throws IOException {
		return ProcessUtils.execute(new String[] { getAbsoluteResourcePath(program) });
	}

	/** Gets the absolute path to a resource file. */
	private static String getAbsoluteResourcePath(String path) {
		return BundleContext.getInstance().getResourceManager().getAbsoluteResourcePath(path);
	}

	/**
	 * Returns the inlined and jitted methods for the testee assembly. All other
	 * trace file content may vary across machines or versions of the .NET
	 * framework, including the number of actually jitted methods (in mscorlib).
	 */
	private static String getNormalizedTraceContent(File traceFile, Set<String> assembliesToCompare)
			throws IOException {
		String content = FileSystemUtils.readFileUTF8(traceFile);

		ListMap<String, String> traceMap = keyValuesMapFor(content, new LoggerMock(), traceFile.getName());

		List<String> inlined = filterMethodInvocationsByAssemblyNumber(traceMap.getCollection(LABEL_INLINED),
				assembliesToCompare);
		List<String> jitted = filterMethodInvocationsByAssemblyNumber(traceMap.getCollection(LABEL_JITTED),
				assembliesToCompare);

		return StringUtils.concat(CollectionUtils.sort(CollectionUtils.unionSet(inlined, jitted)),
				StringUtils.LINE_SEPARATOR);
	}

	/**
	 * Filters invoked methods, keeping only those that contain allowed assembly
	 * keys. Each line in the output corresponds to one method invocation.
	 */
	private static List<String> filterMethodInvocationsByAssemblyNumber(List<String> methodInvocations,
			Set<String> assembliesToCompare) {

		List<String> methodIds = new ArrayList<>();

		for (String methodInvocation : methodInvocations) {
			String assemblyToken = StringUtils.getFirstParts(methodInvocation, 1, KEY_SEPARATOR);

			if (assembliesToCompare.contains(assemblyToken)) {
				methodIds.add(methodInvocation);
			}
		}
		return methodIds;
	}

	/**
	 * Returns "64" if running on an x64 OS and having a .NET framework x64
	 * installed, or "32" in case of an x86 .NET framework.
	 */
	protected static String getBitness() throws IOException {
		return executeProgramInResourceFolder("bitness-checker.exe").getStdout().substring(0, 2);
	}
}

File: Profiler_Test/ProfilerTestBase.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;

namespace Cqse.Teamscale.Profiler.Dotnet
{
    /// <summary>
    /// Base class for testing the .NET Profiler.
    /// </summary>
    public abstract class ProfilerTestBase
    {
        /// <summary>
        /// Label with which jitted methods are prefixed
        /// </summary>
        public const string LABEL_JITTED = "Jitted";

        /// <summary>
        /// Label with which inlined methods are prefixed
        /// </summary>
        public const string LABEL_INLINED = "Inlined";

        /// <summary>
        /// Label with which loaded assemblies are prefixed
        /// </summary>
        public const string LABEL_ASSEMBLY = "Assembly";

        /// <summary>
        /// Separator used to concatenate method keys
        /// </summary>
        public const char KEY_SEPARATOR = ':';

        private List<Process> startedProcesses = new List<Process>();

        /// <summary>
        /// The directory containing profiler solution.
        /// </summary>
        public static DirectoryInfo SolutionRoot => new DirectoryInfo(Path.Combine(TestContext.CurrentContext.TestDirectory, "../../../"));

#if DEBUG

        /// <summary>
        /// Field holding the build configuration, either 'Release' or 'Debug'
        /// </summary>
        protected const string Configuration = "Debug";

#else
        /// <summary>
        /// Field holding the build configuration, either 'Release' or 'Debug'
        /// </summary>
        protected const string Configuration = "Release";
#endif

        /// <summary>
        /// The directory containing the profiler DLLs
        /// </summary>
        protected static readonly string ProfilerDirectory = $"{SolutionRoot}/Profiler/bin/{Configuration}";

        [TearDown]
        public void TearDown()
        {
            startedProcesses.Where(process => !process.HasExited).ToList().ForEach(process => process.Kill());
        }

        /// <summary>
        /// Asserts that the trace file written by the profiler has the same contents as the given reference trace, modulo some normalization.
        /// </summary>
        protected void AssertNormalizedTraceFileEqualsReference(string[] actualTraceContent, int[] assembliesToCompare)
        {
            FileInfo referenceTraceFile = new FileInfo(GetTestDataPath("reference-traces", GetSanitizedTestName() + ".txt"));
            string[] referenceTraceFileContent = File.ReadAllLines(referenceTraceFile.FullName);

            var assemblyIds = new HashSet<int>(assembliesToCompare);
            Assert.AreEqual(ReadNormalizedTraceContent(referenceTraceFileContent, assemblyIds),
                        ReadNormalizedTraceContent(actualTraceContent, assemblyIds),
                        "The normalized contents of the trace files did not match");
        }

        /// <summary>
        /// Returns the absolute path to a test data file.
        /// </summary>
        protected static string GetTestDataPath(params string[] path)
            => Path.Combine(SolutionRoot.FullName, "test-data", Path.Combine(path));

        /// <summary>
        /// An executable file in the TestProgramsDirectory.
        /// </summary>
        protected FileInfo GetTestProgram(string executableName) => new FileInfo(Path.Combine(TestProgramsDirectory.FullName, executableName));

        /// <summary>
        /// The directory containing the testee binaries.
        /// </summary>
        private static DirectoryInfo TestProgramsDirectory =>
            new DirectoryInfo(GetTestDataPath("test-programs"));
        
        /// <summary>
        /// Returns a test-specific directory for temp files
        /// </summary>
        protected static string TestTempDirectory =>
            Path.Combine(SolutionRoot.FullName, "test-tmp", GetSanitizedTestName());
        
        /// <summary>
        /// A temporary directory for a profiler to write trace files to.
        /// </summary>
        protected static DirectoryInfo TestTraceDirectory =>
            new DirectoryInfo(TestTempDirectory).CreateSubdirectory("traces");

        /// <summary>
        /// Creates a unique (and empty) temporary test directory for storing output.
        /// </summary>
        [SetUp]
        protected void CreateTemporaryTestDir()
        {
            var testDir = new DirectoryInfo(TestTempDirectory);
            if (testDir.Exists)
            {
                testDir.Delete(true);
            }

            testDir.Create();
        }

        /// <summary>
        /// Returns a sanatized name for the test case that is valid for paths.
        /// </summary>
        /// <returns></returns>
        private static string GetSanitizedTestName()
        {
            var testDirName = TestContext.CurrentContext.Test.FullName;
            char[] invalidChars = Path.GetInvalidFileNameChars().Union(Path.GetInvalidPathChars()).Distinct().ToArray();
            return string.Join("", testDirName.Split(invalidChars, StringSplitOptions.RemoveEmptyEntries));
        }

        /// <summary>
        /// Returns the inlined and jitted methods for the testee assembly. All other
        /// trace file content may vary across machines or versions of the.NET
        /// framework, including the number of actually jitted methods(in mscorlib).
        /// </summary>
        private static string ReadNormalizedTraceContent(string[] traceFileContent, HashSet<int> assembliesToCompare)
        {
            ILookup<string, string> traceMap = KeyValuesMapFor(traceFileContent);

            IEnumerable<string> inlined = FilterMethodInvocationsByAssemblyNumber(traceMap[LABEL_INLINED], assembliesToCompare);
            IEnumerable<string> jitted = FilterMethodInvocationsByAssemblyNumber(traceMap[LABEL_JITTED], assembliesToCompare);

            var invocations = inlined.Union(jitted).ToList();
            invocations.Sort();

            return string.Join("\\n", invocations);
        }

        /// <summary>
        /// Filters invoked methods, keeping only those that contain allowed assembly keys. Each line in the output corresponds to one method invocation.
        /// </summary>
        private static IEnumerable<string> FilterMethodInvocationsByAssemblyNumber(IEnumerable<string> methodInvocations,
                HashSet<int> assembliesToCompare)
        {
            foreach (string methodInvocation in methodInvocations)
            {
                string assemblyToken = methodInvocation.Split(new char[] { KEY_SEPARATOR }, 2).First();

                if (int.TryParse(assemblyToken, out int assemblyId))
                {
                    if (assembliesToCompare.Contains(int.Parse(assemblyToken)))
                    {
                        yield return methodInvocation;
                    }
                }
            }
        }

        /// <summary>
        /// Parse trace into map from key to list of values.
        /// </summary>
        private static ILookup<string, string> KeyValuesMapFor(string[] coverageReport)
            => coverageReport.Where(line => !string.IsNullOrWhiteSpace(line) && !line.StartsWith("//"))
                .Select(line => line.Split(new char[] { '=' }, 2))
                .ToLookup(split => split[0], split => split[0]);
    }
}
